spot_ros2_control/hardware/src/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,367 @@
+  }
+  return command_interfaces;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_activate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  // command and state should be equal when starting
+  hw_commands_.assign(hw_states_.begin(), hw_states_.end());
+  // Set up the robot using the BD SDK.
+  if (!authenticate_robot(hostname_, username_, password_)) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!start_time_sync()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!check_estop()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!get_lease()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!power_on()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!start_state_stream(std::bind(&StateStreamingHandler::handle_state_streaming, &state_streaming_handler_,
+                                    std::placeholders::_1))) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  stop_state_stream();
+  release_lease();
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_shutdown(const rclcpp_lifecycle::State& /*previous_state*/) {
+  stop_state_stream();
+  release_lease();
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::return_type SpotHardware::read(const rclcpp::Time& /*time*/, const rclcpp::Duration& /*period*/) {
+  const auto& joint_pos = state_streaming_handler_.get_position();
+  const auto& joint_vel = state_streaming_handler_.get_velocity();
+  const auto& joint_load = state_streaming_handler_.get_load();
+  // wait for them to be initialized
+  if (joint_pos.size() == 0 || joint_vel.size() == 0 || joint_load.size() == 0) {
+    return hardware_interface::return_type::OK;
+  }
+  // Ensure that the states received from the Spot SDK will fit into the hw_states_ vector
+  const auto states_size = hw_states_.size();
+  if (interfaces_per_joint_ * joint_pos.size() != states_size ||
+      interfaces_per_joint_ * joint_vel.size() != states_size ||
+      interfaces_per_joint_ * joint_load.size() != states_size) {
+    RCLCPP_FATAL(
+        rclcpp::get_logger("SpotHardware"),
+        "The number of joints and interfaces does not match with the outputted joint states from the Spot SDK!");
+    return hardware_interface::return_type::ERROR;
+  }
+  // Read values into joint states
+  for (uint i = 0; i < states_size; i += interfaces_per_joint_) {
+    hw_states_.at(i) = joint_pos.at(i / interfaces_per_joint_);
+    hw_states_.at(i + 1) = joint_vel.at(i / interfaces_per_joint_);
+    hw_states_.at(i + 2) = joint_load.at(i / interfaces_per_joint_);
+  }
```

Comment:
@khughes-bdai nit^2: nothing against integer long division, but it will say that:
```suggestion
  for (size_t i = 0; i < joint_pos.size(); ++i) {
    hw_states_.at(i * interfaces_per_joint_) = joint_pos.at(i);
    hw_states_.at(i * interfaces_per_joint_ + 1) = joint_vel.at(i);
    hw_states_.at(i * interfaces_per_joint_ + 2) = joint_load.at(i);
  }
```
reads more clearly.